File: Chapter10/Lesson2/Exercise1/Completed/CS/PerfAppInstaller/Class1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using System.Diagnostics;
using System.Configuration.Install;
using System.ComponentModel;
using System.Collections;

namespace PerfAppInstaller
{
    [RunInstaller(true)]
    public class InstallCounter : Installer
    {
        public InstallCounter()
            : base()
        {
        }

        public override void Commit(IDictionary mySavedState)
        {
            base.Commit(mySavedState);
        }

        public override void Install(IDictionary stateSaver)
        {
            base.Install(stateSaver);
            if (!PerformanceCounterCategory.Exists("PerfApp"))
                PerformanceCounterCategory.Create("PerfApp", "Counters for PerfApp",
                    PerformanceCounterCategoryType.SingleInstance,
                    "Clicks", "Times the user has clicked the button.");
        }

        public override void Uninstall(IDictionary savedState)
        {
            base.Uninstall(savedState);
            if (PerformanceCounterCategory.Exists("PerfApp"))
                PerformanceCounterCategory.Delete("PerfApp");
        }

        public override void Rollback(IDictionary savedState)
        {
            base.Rollback(savedState);
            if (PerformanceCounterCategory.Exists("PerfApp"))
                PerformanceCounterCategory.Delete("PerfApp");
        }
    }
}

File: Chapter10/Lesson2/Exercise1/Completed/VB/PerfAppInstaller/Class1.vb
Imports System.Configuration.Install
Imports System.ComponentModel

<RunInstaller(True)> _
        Public Class InstallCounter
    Inherits Installer
    Public Sub New()
        MyBase.New()
    End Sub

    Public Overloads Overrides Sub Commit(ByVal mySavedState As IDictionary)
        MyBase.Commit(mySavedState)
    End Sub

    Public Overloads Overrides Sub Install(ByVal stateSaver As IDictionary)
        MyBase.Install(stateSaver)
        If Not PerformanceCounterCategory.Exists("PerfApp") Then
            PerformanceCounterCategory.Create("PerfApp", "Counters for PerfApp", PerformanceCounterCategoryType.SingleInstance, "Clicks", "Times the user has clicked the button.")
        End If
    End Sub

    Public Overloads Overrides Sub Uninstall(ByVal savedState As IDictionary)
        MyBase.Uninstall(savedState)
        If PerformanceCounterCategory.Exists("PerfApp") Then
            PerformanceCounterCategory.Delete("PerfApp")
        End If
    End Sub

    Public Overloads Overrides Sub Rollback(ByVal savedState As IDictionary)
        MyBase.Rollback(savedState)
        If PerformanceCounterCategory.Exists("PerfApp") Then
            PerformanceCounterCategory.Delete("PerfApp")
        End If
    End Sub
End Class
